wpimath/src/test/java/edu/wpi/first/math/autodiff/GradientTest.java

```diff
@@ -0,0 +1,591 @@
+// the WPILib BSD license file in the root directory of this project.
+
+package edu.wpi.first.math.autodiff;
+
+import static edu.wpi.first.math.MatrixAssertions.assertEquals;
+import static edu.wpi.first.math.autodiff.Variable.abs;
+import static edu.wpi.first.math.autodiff.Variable.acos;
+import static edu.wpi.first.math.autodiff.Variable.asin;
+import static edu.wpi.first.math.autodiff.Variable.atan;
+import static edu.wpi.first.math.autodiff.Variable.atan2;
+import static edu.wpi.first.math.autodiff.Variable.cbrt;
+import static edu.wpi.first.math.autodiff.Variable.cos;
+import static edu.wpi.first.math.autodiff.Variable.cosh;
+import static edu.wpi.first.math.autodiff.Variable.exp;
+import static edu.wpi.first.math.autodiff.Variable.hypot;
+import static edu.wpi.first.math.autodiff.Variable.log;
+import static edu.wpi.first.math.autodiff.Variable.log10;
+import static edu.wpi.first.math.autodiff.Variable.pow;
+import static edu.wpi.first.math.autodiff.Variable.signum;
+import static edu.wpi.first.math.autodiff.Variable.sin;
+import static edu.wpi.first.math.autodiff.Variable.sinh;
+import static edu.wpi.first.math.autodiff.Variable.sqrt;
+import static edu.wpi.first.math.autodiff.Variable.tan;
+import static edu.wpi.first.math.autodiff.Variable.tanh;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.ejml.simple.SimpleMatrix;
+import org.junit.jupiter.api.Test;
+
+class GradientTest {
+  @Test
+  void testTrivialCase() {
+    var a = new Variable();
+    a.setValue(10);
+    var b = new Variable();
+    b.setValue(20);
+    var c = a;
+
+    assertEquals(1.0, new Gradient(a, a).value().get(0, 0));
+    assertEquals(0.0, new Gradient(a, b).value().get(0, 0));
+    assertEquals(1.0, new Gradient(c, a).value().get(0, 0));
+    assertEquals(0.0, new Gradient(c, b).value().get(0, 0));
+  }
+
+  @Test
+  void testUnaryPlus() {
+    var a = new Variable();
+    a.setValue(10);
+    var c = a.unaryPlus();
+
+    assertEquals(a.value(), c.value());
+    assertEquals(1.0, new Gradient(c, a).value().get(0, 0));
+  }
+
+  @Test
+  void testUnaryMinus() {
+    var a = new Variable();
+    a.setValue(10);
+    var c = a.unaryMinus();
+
+    assertEquals(a.unaryMinus().value(), c.value());
+    assertEquals(-1.0, new Gradient(c, a).value().get(0, 0));
+  }
+
+  @Test
+  void testIdenticalVariables() {
+    var a = new Variable();
+    a.setValue(10);
+    var x = a;
+    var c = a.times(a).plus(x);
+
+    assertEquals(a.value() * a.value() + x.value(), c.value());
+    assertEquals(
+        2 * a.value() + new Gradient(x, a).value().get(0, 0), new Gradient(c, a).value().get(0, 0));
+    assertEquals(
+        2 * a.value() * new Gradient(x, a).value().get(0, 0) + 1,
+        new Gradient(c, a).value().get(0, 0));
+  }
+
+  @Test
+  void testElementary() {
+    var a = new Variable();
+    a.setValue(1.0);
+    var b = new Variable();
+    b.setValue(2.0);
+    var c = new Variable();
+    c.setValue(3.0);
+
+    c = a.times(-2);
```

Comment:
Why do we create `c`, set the value, and then immediately replace the reference?

Reply:
The test makes more sense in C++ where there's actually an implementation to test.